### PR TITLE
Add 'install --formula', rewrite '--deps-only'

### PR DESCRIPTION
**Pull request by AltGr**

### CODE REVIEW

**Comment by AltGr** [comment]
Should be finished, just needs some git history cleanup + tests

**Comment by rjbou** [comment]
test the test from #4914

File: src/client/opamClient.ml [hunk 1]
```diff
@@ -941,6 +959,72 @@ let install_t t ?ask ?(ignore_conflicts=false) ?(depext_only=false)
...
+          in
+          let dname = nodup 2 @@ of_string ("deps-of-" ^ to_string name) in
+          OpamPackage.Name.Map.add name dname)
+        names OpamPackage.Name.Map.empty
+    else OpamPackage.Name.Map.empty
+  in
+  let t =
+    (* add deps-of-xxx packages to replace each atom *)
+    OpamPackage.Name.Map.fold (fun name dname t ->
+        let ats = List.filter (fun (n,_) -> n = name) atoms in
```
Comment (kit-ty-kate) [inline src/client/opamClient.ml:984]:
```suggestion
        let ats = List.filter (fun (n,_) -> OpamPackage.Name.equal n name) atoms in
```

File: src/client/opamClient.ml [hunk 2]
```diff
@@ -1004,48 +1088,29 @@ let install_t t ?ask ?(ignore_conflicts=false) ?(depext_only=false)
...
   in
-  let request = OpamSolver.request ~install:atoms () in
+  let request =
+    OpamSolver.request ()
+      ~install:(atoms @ deps_atoms)
+      ~deprequest:(OpamFormula.to_atom_formula formula)
+  in
+  let requested =
+    OpamPackage.Name.Set.of_list (List.rev_map fst (atoms @ deps_atoms))
+  in
```
Comment (kit-ty-kate) [inline src/client/opamClient.ml:1109]:
```suggestion
  let requested = atoms @ deps_atoms in
  let request =
    OpamSolver.request ()
      ~install:requested
      ~deprequest:(OpamFormula.to_atom_formula formula)
  in
  let requested =
    OpamPackage.Name.Set.of_list (List.rev_map fst requested)
  in
```

File: src/solver/opamCudf.ml [hunk 3]
```diff
@@ -31,9 +31,20 @@ let opam_invariant_package_version = 1
...
+let opam_deprequest_package_version = 1
+
+let opam_deprequest_package =
+  opam_deprequest_package_name, opam_deprequest_package_version
+
 let is_opam_invariant p =
   p.Cudf.package = opam_invariant_package_name
 
+let is_opam_deprequest p =
+  p.Cudf.package = opam_deprequest_package_name
```
Comment (kit-ty-kate) [inline src/solver/opamCudf.ml:46]:
```suggestion
  String.equal p.Cudf.package opam_invariant_package_name

let is_opam_deprequest p =
  String.equal p.Cudf.package opam_deprequest_package_name
```

File: src/solver/opamSolver.ml [hunk 4]
```diff
@@ -482,17 +452,26 @@ let resolve universe request =
     in
     { request with extra_attributes }
   in
-  let request = cleanup_request universe request in
+  let request, deprequest_pkg =
+    let conj = OpamFormula.ands_to_list request.wish_install in
+    let conj, deprequest =
+      List.partition (function Atom _ -> true | _ -> false) conj
```
Comment (kit-ty-kate) [inline src/solver/opamSolver.ml:458]:
```suggestion
      List.partition (function Atom _ -> true | Empty | Block _ | And _ | Or _ -> false) conj
```

File: tests/reftests/install-formula.test [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  - install ocamlfind       1.8.0          [required by ocp-build]
+  - install base-ocamlbuild base           [required by ocamlbuild]
+  - install ocp-build       1.99.19-beta   [required by ocp-index]
+  - install ocamlbuild      0              [required by cmdliner]
+  - install cmdliner        0.9.7          [required by ocp-index]
+  - install ocp-indent      1.4.0          [required by ocp-index]
+  - install ocp-index       1.0.0
+===== 14 to install =====
+### OPAMNODEPEXTS=1
+### opam install genet --formula '["mysql" ("conf-mariadb" & "mariadb" | "conf-postgresql")]'
```
Comment (kit-ty-kate) [inline tests/reftests/install-formula.test:157]: what’s with the brackets here? I’ve never seen this syntax before and it’s not in the documentation

**Review by kit-ty-kate — APPROVED** [review]
Apart from the minor bracket interogation above LGTM.

File: tests/reftests/install-formula.test (hunk 5 above)
Comment (AltGr) [reply tests/reftests/install-formula.test:157]: it's just the same syntax as the `depends:` field

Comment (kit-ty-kate) [reply tests/reftests/install-formula.test:157]: ah i see

**Review by rjbou — APPROVED** [review]
all good! thanks!

**Comment by rjbou** [comment]
rebased

**Comment by rjbou** [comment]
closes #4914

**Pull request merged**